onnxruntime/core/session/inference_session.cc

```diff
@@ -731,6 +767,99 @@ common::Status InferenceSession::TransformGraph(onnxruntime::Graph& graph,
+
+  if (!bytes_stream) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                           "Load model from ", ToMBString(model_uri), " failed. Only ",
+                           bytes_stream.gcount(), "/", num_bytes, " bytes were able to be read.");
+  }
+
+  return Status::OK();
+}
+
+Status InferenceSession::LoadOrtModel(const std::string& model_uri) {
+  return LoadOrtModel(
+      [&](gsl::span<const uint8_t>& bytes) {
+        ORT_RETURN_IF_ERROR(LoadOrtModelBytes(model_uri, model_location_,
+                                              ort_format_model_num_bytes_, ort_format_model_bytes_));
+
+        bytes = gsl::make_span<const uint8_t>(ort_format_model_bytes_.get(), ort_format_model_num_bytes_);
+        return Status::OK();
+      });
+}
+
+#ifdef WIN32
+Status InferenceSession::LoadOrtModel(const std::wstring& model_uri) {
+  return LoadOrtModel(
+      [&](gsl::span<const uint8_t>& bytes) {
+        ORT_RETURN_IF_ERROR(LoadOrtModelBytes(model_uri, model_location_,
+                                              ort_format_model_num_bytes_, ort_format_model_bytes_));
+
+        bytes = gsl::make_span<const uint8_t>(ort_format_model_bytes_.get(), ort_format_model_num_bytes_);
+        return Status::OK();
+      });
+}
+#endif
+
+Status InferenceSession::LoadOrtModel(const void* model_data, int model_data_len) {
+  return LoadOrtModel([&](gsl::span<const uint8_t>& bytes) {
+    bytes = gsl::make_span<const uint8_t>(static_cast<const uint8_t*>(model_data), model_data_len);
+    return Status::OK();
+  });
+}
+
+Status InferenceSession::LoadOrtModel(std::function<Status(gsl::span<const uint8_t>&)> get_serialized_bytes) {
+  static_assert(FLATBUFFERS_LITTLEENDIAN, "ORT format only supports little-endian machines");
+
+  std::lock_guard<onnxruntime::OrtMutex> l(session_mutex_);
+
+  if (is_model_loaded_) {  // already loaded
+    Status status(common::ONNXRUNTIME, common::MODEL_LOADED, "This session already contains a loaded model.");
+    LOGS(*session_logger_, ERROR) << status.ErrorMessage();
+    return status;
+  }
+
+  if (is_inited_) {
+    Status status(common::ONNXRUNTIME, common::MODEL_LOADED, "This session has already been initialized.");
+    LOGS(*session_logger_, ERROR) << status.ErrorMessage();
+    return status;
+  }
+
+  gsl::span<const uint8_t> bytes;
+  ORT_RETURN_IF_ERROR(get_serialized_bytes(bytes));
+
+  const auto* fbs_session = fbs::GetInferenceSession(bytes.data());
+  ORT_RETURN_IF_NOT(nullptr != fbs_session, "Serialized ORT model is missing InferenceSession");
+
+  const auto* fbs_model = fbs_session->model();
+  ORT_RETURN_IF_NOT(nullptr != fbs_model, "Serialized ORT model is missing Model instance");
+
+  // need to go from unique_ptr to shared_ptr when moving into model_
+  std::unique_ptr<Model> tmp_model;
+  ORT_RETURN_IF_ERROR(Model::LoadFromOrtFormat(*fbs_model, *session_logger_, tmp_model));
+  ORT_RETURN_IF_ERROR(SaveModelMetadata(*tmp_model));
```

Comment:
Should something be checking the version that was used to serialize the data? We could still attempt to deserialize in a future version (optional), but as we promise no backwards compatibility in the first version as it's experimental we should at least have a loud warning saying things may break due to a version mismatch.